pkg/shed/db.go

```diff
@@ -74,83 +88,361 @@ func NewDB(path string, logger logging.Logger) (db *DB, err error) {
+	} else {
+		db.logger.Tracef("got value with len : %d for key : %s", len(value), string(key))
+		db.metrics.GetCount.Inc()
 	}
-	db.metrics.GetCounter.Inc()
-	return value, nil
+
+	return value, err
 }
 
-// Has wraps LevelDB Has method to increment metrics counter.
+// Has checks if the given key is present in the database.
+// it returns a bool indicating true or false OR error if it encounters one during the operation.
 func (db *DB) Has(key []byte) (yes bool, err error) {
-	yes, err = db.ldb.Has(key, nil)
+	db.logger.Tracef("Has value with key %s", string(key))
+	yes = false
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		item, err := txn.Get(key)
+		if err != nil {
+			if err == badger.ErrKeyNotFound {
+				return ErrNotFound
+			}
+			return err
+		}
+		return item.Value(func(val []byte) error {
+			yes = true
+			db.metrics.HasCount.Inc()
+			return nil
+		})
+	})
 	if err != nil {
-		db.logger.Debugf("encountered error during HAS of key %s. Error: %s ", string(key), err.Error())
-		db.metrics.HasFailCounter.Inc()
-		return false, err
+		db.logger.Errorf("key %s not found in DB", string(key))
+		db.logger.Debugf("key %s not found in DB. Error: %v", string(key), err.Error())
+		db.metrics.HasFailCount.Inc()
 	}
-	db.metrics.HasCounter.Inc()
+
+	if yes {
+		db.logger.Tracef("found key %s", string(key))
+	} else {
+		db.logger.Tracef("could not found key %s", string(key))
+	}
+
 	return yes, nil
 }
 
-// Delete wraps LevelDB Delete method to increment metrics counter.
+// Delete removed the key and value if a given key is present in the DB.
 func (db *DB) Delete(key []byte) (err error) {
-	err = db.ldb.Delete(key, nil)
-	if err != nil {
-		db.logger.Debugf("could not DELETE key %s. Error: %s ", string(key), err.Error())
-		db.metrics.DeleteFailCounter.Inc()
+	db.logger.Tracef("deleting key %s along with value", string(key))
+	return db.bdb.Update(func(txn *badger.Txn) (err error) {
+		db.metrics.DeleteCount.Inc()
+		err = txn.Delete(key)
+		if err != nil {
+			db.logger.Errorf("could not delete key %s from DB", string(key))
+			db.logger.Debugf("could not delete key %s from DB. Error: %v", string(key), err.Error())
+			db.metrics.DeleteFailCount.Inc()
+		} else {
+			db.logger.Tracef("deleted key %s", string(key))
+		}
 		return err
+	})
+}
+
+// Count gives a count of all the keys present in the DB.
+func (db *DB) Count(ctx context.Context) (count int, err error) {
+	db.logger.Tracef("counting all keys")
+	db.metrics.TotalCount.Inc()
+	err = db.bdb.View(func(txn *badger.Txn) (err error) {
+		o := badger.DefaultIteratorOptions
+		o.PrefetchValues = false
+		i := txn.NewIterator(o)
+		defer i.Close()
+		for i.Rewind(); i.Valid(); i.Next() {
+			item := i.Item()
+			k := item.KeySize()
+			if k < 1 {
+				continue
+			}
+			count++
+		}
+		return nil
+	})
+	if err != nil {
+		db.logger.Errorf("error while counting keys in DB")
+		db.logger.Debugf("error while doing count. Error : %w", err.Error())
+		db.metrics.TotalFailCount.Inc()
```

Comment:
Should this block return and avoid the Trace log line below?

Reply:
no. I want ed to print the count if trace is enabled.

Reply:
Should the count be irrelevant if there is an error?

Reply:
ok. Add this in an else